### PR TITLE
Build listener to backfill position

### PR DESCRIPTION
**Pull request by thomtrp**
We want the position set for all our records.
Currently, this is still possible to create records without a position through the rest API.

This PR:
- adds a listener that listens on all record creations and trigger a job if the position is not set
- a job that simply call a new record position backfill service
- the service that backfill the position

I re-used the existing logic that overrides the position in the args during a normal creation through web. I put it in a new factory `RecordPositionFactory`.

TODO:
- testing once the strategy is validated

### CODE REVIEW

File: packages/twenty-server/src/workspace/workspace-query-runner/factories/record-position.factory.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+} from 'src/workspace/workspace-query-builder/factories/record-position-query.factory';
+
+@Injectable()
+export class RecordPositionFactory {
+  constructor(
+    private readonly workspaceDataSourceService: WorkspaceDataSourceService,
+    private readonly recordPositionQueryFactory: RecordPositionQueryFactory,
+  ) {}
+
+  async create(
```
Comment (thomtrp) [inline packages/twenty-server/src/workspace/workspace-query-runner/factories/record-position.factory.ts:16]: I simply moved the logic of `buildPositionValue` in there

File: packages/twenty-server/src/workspace/workspace-query-runner/workspace-query-runner.module.ts [hunk 2]
```diff
@@ -13,7 +14,11 @@ import { WorkspaceQueryRunnerService } from './workspace-query-runner.service';
     WorkspaceDataSourceModule,
     WorkspacePreQueryHookModule,
   ],
-  providers: [WorkspaceQueryRunnerService, ...workspaceQueryRunnerFactories],
+  providers: [
+    WorkspaceQueryRunnerService,
+    ...workspaceQueryRunnerFactories,
+    RecordPositionListener,
```
Comment (thomtrp) [inline packages/twenty-server/src/workspace/workspace-query-runner/workspace-query-runner.module.ts:20]: @charlesBochet @magrinj I don't know if the service / job / factory / listener should live into the query runner folder. Is there a better place ?

File: packages/twenty-server/src/workspace/workspace-query-builder/factories/record-position-query.factory.ts [hunk 3]
```diff
@@ -1,21 +1,74 @@
...
 export class RecordPositionQueryFactory {
   async create(
+    recordPositionQueryType: RecordPositionQueryType,
+    positionValue: 'first' | 'last' | number,
+    objectMetadata: { isCustom: boolean; nameSingular: string },
+    dataSourceSchema: string,
+    recordId?: string,
+  ): Promise<string> {
+    const name =
+      (objectMetadata.isCustom ? '_' : '') + objectMetadata.nameSingular;
```
Comment (Weiko) [inline packages/twenty-server/src/workspace/workspace-query-builder/factories/record-position-query.factory.ts:18]: nit: We should probably have a helper for that for when we will remove this logic

File: packages/twenty-server/src/workspace/workspace-query-builder/factories/record-position-query.factory.ts [hunk 4]
```diff
@@ -1,21 +1,74 @@
...
             WHERE "position" IS NOT NULL ORDER BY "position" ${orderByDirection} LIMIT 1`;
   }
+
+  private async createForUpdate(
+    positionValue: number,
+    name: string,
+    dataSourceSchema: string,
+    recordId: string,
+  ): Promise<string> {
+    return `UPDATE ${dataSourceSchema}."${name}"
```
Comment (Weiko) [inline packages/twenty-server/src/workspace/workspace-query-builder/factories/record-position-query.factory.ts:70]: prepared statements would be safer

File: packages/twenty-server/src/workspace/workspace-query-runner/jobs/record-position-backfill.job.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+@Injectable()
+export class RecordPositionBackfillJob
+  implements MessageQueueJob<RecordPositionBackfillJobData>
+{
+  constructor(
+    private readonly recordPositionBackfillService: RecordPositionBackfillService,
+  ) {}
+
+  async handle(data: RecordPositionBackfillJobData): Promise<void> {
+    this.recordPositionBackfillService.backfill(
```
Comment (Weiko) [inline packages/twenty-server/src/workspace/workspace-query-runner/jobs/record-position-backfill.job.ts:22]: missing await?

File: packages/twenty-server/src/integrations/event-emitter/types/object-record-create.event.ts [hunk 6]
```diff
@@ -1,6 +1,12 @@
...
 
+export type CreatedObjectMetadata = {
+  nameSingular: string;
+  isCustom: boolean;
+};
+
 export class ObjectRecordCreateEvent<T extends BaseObjectMetadata> {
   workspaceId: string;
   createdRecord: T;
+  createdObjectMetadata: CreatedObjectMetadata;
```
Comment (Weiko) [inline packages/twenty-server/src/integrations/event-emitter/types/object-record-create.event.ts:11]: Why do you need this? Don't you have what you need inside createdRecord?

**Review by Weiko — COMMENTED** [review]
Left a few comments @thomtrp

File: packages/twenty-server/src/workspace/workspace-query-runner/jobs/record-position-backfill.job.ts (hunk 5 above)
Comment (thomtrp) [reply packages/twenty-server/src/workspace/workspace-query-runner/jobs/record-position-backfill.job.ts:22]: why so? I do not need to return the value anyway so why waiting the result?

File: packages/twenty-server/src/integrations/event-emitter/types/object-record-create.event.ts (hunk 6 above)
Comment (thomtrp) [reply packages/twenty-server/src/integrations/event-emitter/types/object-record-create.event.ts:11]: No, the record contains only the data of the value, not the metadata

Comment (Weiko) [reply packages/twenty-server/src/integrations/event-emitter/types/object-record-create.event.ts:11]: Ah yes!

**Review by Weiko — APPROVED** [review]
LGTM

**Pull request merged**